File: src/converter/plugins/convertVoid.ts
import ts, {Node} from "typescript";
import {createSimplePlugin} from "../plugin.js";

function extractReturnType(node: Node) {
    // signatures
    if (ts.isCallSignatureDeclaration(node)) return node.type
    if (ts.isConstructSignatureDeclaration(node)) return node.type
    if (ts.isMethodSignature(node)) return node.type

    // function like
    if (ts.isFunctionDeclaration(node)) return node.type
    if (ts.isMethodDeclaration(node)) return node.type

    // types
    if (ts.isFunctionTypeNode(node)) return node.type
    if (ts.isConstructorTypeNode(node)) return node.type

    return undefined
}

export const convertVoid = createSimplePlugin(node => {
    if (node.kind !== ts.SyntaxKind.VoidKeyword) return null

    const returnType = extractReturnType(node.parent)

    if (returnType !== undefined) {
        return "Unit"
    }

    return "js.core.Void"
})

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertVoid.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.createSimplePlugin
import typescript.*

private fun extractReturnType(node: Node): Node? {
    // signatures
    if (isCallSignatureDeclaration(node)) return node.type
    if (isConstructSignatureDeclaration(node)) return node.type
    if (isMethodSignature(node)) return node.type

    // function like
    if (isFunctionDeclaration(node)) return node.type
    if (isMethodDeclaration(node)) return node.type

    // types
    if (isFunctionTypeNode(node)) return node.type
    if (isConstructorTypeNode(node)) return node.type

    return null
}

val convertVoid = createSimplePlugin plugin@{ node, _, _ ->
    if (node.kind !== SyntaxKind.VoidKeyword) return@plugin null

    val returnType = extractReturnType(node.parent)

    if (returnType != null) {
        return@plugin "Unit"
    }

    "js.core.Void"
}
